src/Server/RestRouterHandlers/SearchHandler.cpp

```diff
@@ -176,4 +178,66 @@ std::shared_ptr<WriteBufferFromHTTPServerResponse> SearchHandler::getOutputBuffe
 
     return out;
 }
+
+void SearchHandler::setQuerySettings() const
+{
+    static const NameSet reserved_param_names{
+        "compress",
+        "decompress",
+        "user",
+        "password",
+        "quota_key",
+        "query_id",
+        "stacktrace",
+        "buffer_size",
+        "wait_end_of_query",
+        "session_id",
+        "session_timeout",
+        "session_check"};
+
+    auto param_could_be_skipped = [&](const String & name) {
+        /// Empty parameter appears when URL like ?&a=b or a=b&&c=d. Just skip them for user's convenience.
+        if (name.empty())
+            return true;
+
+        if (reserved_param_names.count(name))
+            return true;
+
+        return false;
+    };
+
+    const auto & query_id = getQueryParameter("x-bdg-request-id");
+    if (!query_id.empty())
+        query_context->setCurrentQueryId(query_id);
+
+    /// FIXME to support cascaded write buffer and session
+    SettingsChanges settings_changes;
+    String default_format;
+    for (const auto & [key, value] : *query_parameters)
+    {
+        if (param_could_be_skipped(key))
+            continue;
+
+        if (key == "default_format")
+            default_format = value;
+        else
+        {
+            /// Other than query parameters are treated as settings.
+            if (startsWith(key, "param_"))
+            {
+                /// Save name and values of substitution in dictionary.
+                const String parameter_name = key.substr(strlen("param_"));
+                settings_changes.push_back({parameter_name, value});
+            }
+            else
+                settings_changes.push_back({key, value});
+        }
+    }
+
+    if (!default_format.empty())
+        query_context->setDefaultFormat(default_format);
```

Comment:
If we set a different format, do we still have `query_id` etc in the response or it is completely ClickHouse response format ? I was thinking we shall have some consistent high level response format for search API, something like 
```
{
    "query_id" : <query_id>,
    "data" : <data>,
    "format" : <format>,
}
```

Reply:
This is why we choose 'JSONXXX' format by default. otherwise it is hard to keep the consistant output format. Actually the response of search api is just the raw data.  query_id is set by header in the same way like native HTTPHandler. The reason is performance. As search api is a stream api, if I reformat the response, I need to buffer the whole search results and what is worse, the client cannot get any response until the search completes. That is the reason why I disable send_progress_in_header neither.

Reply:
If we use JSONCompact, frontend need wait until the whole JSON object fully transmitted to frontend which is in general good for different programming clients as it is very easy for them to parse and in most cases we like this behavior and I think by default we can limit the returned data set for this format since it doesn't make too much sense to me to stream 1 million events to frontend for example.

For large data set export case,  we probably like to use the other format (JSONXXXXEachLine) because in that scenario, it is assumed a long run background job continuously exporting data.

Essentially, what i was proposing here is in normal analytic / ad-hoc query case, we use our JSON format with query ID etc consistently with other APIs and limit the returned data set . For background data export case, i was struggling if need do this via a REST API, but providing one seems not making any harm.

Reply:
The other reason why I like to limit the returned data set by default is resource consumption.

Reply:
Use a new JSON like output format in future instead. it can be done in another PR.